src/background/dev-tools-monitor.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { OnDevToolStatusPayload } from 'background/actions/action-payloads';
+import { TabContextManager } from 'background/tab-context-manager';
+import { BrowserAdapter } from 'common/browser-adapters/browser-adapter';
+import { Messages } from 'common/messages';
+import { PromiseFactory, TimeoutError } from 'common/promises/promise-factory';
+import { DevToolsStatusResponse } from 'common/types/dev-tools-messages';
+import _ from 'lodash';
+
+export class DevToolsMonitor {
+    constructor(
+        private readonly browserAdapter: BrowserAdapter,
+        private readonly promiseFactory: PromiseFactory,
+        protected readonly activeDevtoolTabIds: number[],
+        private readonly tabContextManager: TabContextManager,
+        private readonly messageTimeoutMilliseconds = 5000,
+        private readonly pollIntervalMilliseconds = 1000,
+    ) {}
+
+    public startMonitoringDevtool(tabId: number): void {
+        this.addActiveDevtool(tabId);
+
+        if (this.activeDevtoolTabIds.length === 1) {
+            // Do not await, we want to continue running other synchronous code
+            this.monitorActiveDevtools();
+        }
+    }
+
+    private addActiveDevtool(tabId: number): void {
+        if (!_.isNil(tabId) && !this.activeDevtoolTabIds.includes(tabId)) {
+            this.activeDevtoolTabIds.push(tabId);
+        }
+    }
+
+    private removeActiveDevtool(tabId: number): void {
+        const index = this.activeDevtoolTabIds.indexOf(tabId);
+        if (index >= 0) {
+            this.activeDevtoolTabIds.splice(index, 1);
+        }
+    }
+
+    protected async monitorActiveDevtools(): Promise<void> {
```

Comment:
Why does this need to be protected? I don't see any classes extending this one, is that something that will be added in the future?

Reply:
This function is overwritten to be public in a test class (see comment below)